File: julia/test/typed_source_location_contract_test.jl
# FUTURE-PARITY-BACKLOG.14.2.4.0.2 — dormant Julia typed source-location RED.
#
# Ordinary Julia discovery is the explicit include list in `test/runtests.jl`;
# that list deliberately omits this pre-admission consumer. Run either mode
# through repository-local project data from the repository root:
#
#   LINKEDSPEC_JULIA_TYPED_SOURCE_RED_MODE=core \
#     bash tools/run_julia_project_data.sh --project=julia \
#       --startup-file=no --history-file=no -e \
#       'using LinkedSpecJulia, JSON3, Test; include("julia/test/typed_source_location_contract_test.jl")'
#
#   LINKEDSPEC_JULIA_TYPED_SOURCE_RED_MODE=projection \
#     bash tools/run_julia_project_data.sh --project=julia \
#       --startup-file=no --history-file=no -e \
#       'using LinkedSpecJulia, JSON3, Test; include("julia/test/typed_source_location_contract_test.jl")'
#
# Core implementation makes `core` green. Projection implementation then
# makes the independently nested `projection` mode green. Admission removes
# this mode switch and adds the unchanged assertions to `test/runtests.jl`.

using LinkedSpecJulia
using JSON3
using Test

const JULIA_TYPED_SOURCE_RED_MODE = get(
    ENV,
    "LINKEDSPEC_JULIA_TYPED_SOURCE_RED_MODE",
    "core",
)

if !(JULIA_TYPED_SOURCE_RED_MODE in ("core", "projection"))
    error(
        "LINKEDSPEC_JULIA_TYPED_SOURCE_RED_MODE must be 'core' or 'projection', " *
        "got '$(JULIA_TYPED_SOURCE_RED_MODE)'",
    )
end

# The namespace is deliberately private. LinkedSpecJulia already exports an
# unrelated parser `SourceSpan`; nesting the neutral Position/Span vocabulary
# prevents a collision and does not create authored DSL or facade methods.
const JuliaTypedSource = getproperty(LinkedSpecJulia, :SourceLocation)

# Projection lookup is ordered strictly after the core namespace. Today both
# modes stop at the missing core. Once the core lands, projection mode advances
# to one exact missing projection API instead of reclassifying that failure.
const JULIA_TYPED_SOURCE_PROJECTION_ROWS = if JULIA_TYPED_SOURCE_RED_MODE == "projection"
    getproperty(LinkedSpecJulia, :typed_source_projection_rows)
else
    nothing
end
const JULIA_TYPED_SOURCE_COMPATIBILITY_ALIASES = if JULIA_TYPED_SOURCE_RED_MODE == "projection"
    getproperty(LinkedSpecJulia, :typed_source_compatibility_aliases)
else
    nothing
end

const JULIA_TYPED_SOURCE_CONTRACT = JSON3.read(
    read(
        normpath(
            joinpath(
                @__DIR__,
                "..",
                "..",
                "capability_conformance",
                "typed_source_location_contract.json",
            ),
        ),
        String,
    ),
    Dict{String,Any},
)

const JULIA_TYPED_SOURCE_CONTEXT = JuliaTypedSource.SourceLocationContext(
    rule_role = "typed_source_fixture_rule",
    invocation_role = "typed_source_fixture_invocation",
)

const JULIA_TYPED_SOURCE_CURSOR_SOURCE = """Top::
 /ab/ -> Done {
  after_match = cursor_pos()
  save_cursor()
  rewind_match_start()
  match_start = cursor_pos()
  restore_cursor()
  restored_match = cursor_pos()
  save_cursor()
  rewind_entry_start()
  entry_start = cursor_pos()
  restore_cursor()
  restored_entry = cursor_pos()
  return(hash(
   "after_match", after_match,
   "match_start", match_start,
   "restored_match", restored_match,
   "entry_start", entry_start,
   "restored_entry", restored_entry
  ))
 }

Done:
 /ab/
"""

const JULIA_TYPED_SOURCE_CURSOR_EXPECTED = Dict{String,Any}(
    "after_match" => 2,
    "entry_start" => 0,
    "match_start" => 0,
    "restored_entry" => 2,
    "restored_match" => 2,
)

const JULIA_TYPED_SOURCE_ALIAS_SOURCE = """Top::OR{1,1}
 /(?<name>ab)/
 I { started = capture_slice_here() }
 E {
  return(array(
   started,
   capture_from_rule_start(),
   capture_len_from_rule_start(),
   capture_slice_length(),
   capture_rest_length(),
   entry_named_map(),
   match_named_map()
  ))
 }
"""

const JULIA_TYPED_SOURCE_ALIAS_EXPECTED = Any[
    nothing,
    "é🙂  ",
    4,
    4,
    6,
    Dict{String,Any}("name" => "ab"),
    Dict{String,Any}("name" => "ab"),
]

function _julia_typed_source_object_rows(key::AbstractString)
    return JULIA_TYPED_SOURCE_CONTRACT[String(key)]
end

function _julia_typed_source_decoded_sources()
    return Dict{String,String}(
        String(fixture["id"]) => String(fixture["decoded_text"])
        for fixture in _julia_typed_source_object_rows("sources")
    )
end

function _julia_typed_source_position(authority, source_id::AbstractString, offset::Int)
    return JuliaTypedSource.position(
        authority;
        source_id = String(source_id),
        offset,
        context = JULIA_TYPED_SOURCE_CONTEXT,
    )
end

function _julia_typed_source_expect_diagnostic(
    diagnostic_id::AbstractString,
    operation::Function,
    expected_context::Dict{String,Any},
)
    captured = try
        operation()
        nothing
    catch error
        error
    end
    @test captured isa JuliaTypedSource.SourceLocationException
    captured isa JuliaTypedSource.SourceLocationException || return

    diagnostic = only(
        row for row in _julia_typed_source_object_rows("diagnostics")
        if row["id"] == diagnostic_id
    )
    record = JuliaTypedSource.to_json(captured)
    @test record["code"] == diagnostic["code"]
    @test record["phase"] == diagnostic["phase"]
    for field in diagnostic["required_context"]
        @test haskey(record, String(field))
    end
    for (key, value) in expected_context
        @test record[key] == value
    end
    for forbidden in (
        "decoded_text",
        "source_text",
        "path",
        "match",
        "parser_state",
        "host_reference",
    )
        @test !haskey(record, forbidden)
    end
end

function _julia_typed_source_compile(source::AbstractString)
    parsed = parse_spec(source)
    validate_spec(parsed)
    return parsed, compile_spec(parsed)
end

function _julia_typed_source_assert_carriers(
    source::AbstractString,
    input::AbstractString,
    identity::AbstractString,
    expected,
)
    parsed, compiled = _julia_typed_source_compile(source)
    native = runtime_parse(LinkedSpecRuntimeEngine(compiled), input).value

    normalized = JSON3.read(JSON3.write(to_json(parsed)), Dict{String,Any})
    reconstructed_spec = from_json(SpecFile, normalized)
    validate_spec(reconstructed_spec)
    reconstructed = runtime_parse(
        LinkedSpecRuntimeEngine(compile_spec(reconstructed_spec)),
        input,
    ).value

    generated = execute_generated_parser_v2(
        compiled,
        build_generated_rule_plan(compiled),
        input,
        identity,
    )

    @test native == expected
    @test reconstructed == expected
    @test generated == expected
end

@testset "Julia dormant typed source-location contract" begin
    @testset "neutral immutable values and coordinate conversions" begin
        @test JULIA_TYPED_SOURCE_CONTRACT["contract_id"] ==
              "linkedspec-typed-source-location-v1"
        expected_counts = JULIA_TYPED_SOURCE_CONTRACT["expected_counts"]
        @test expected_counts["sources"] == 3
        @test expected_counts["position_conversions"] == 7
        @test expected_counts["direct_spans"] == 6
        @test expected_counts["derived_text_cases"] == 3

        sources = _julia_typed_source_decoded_sources()
        authority = JuliaTypedSource.SourceAuthority(sources = sources)
        positions = Dict{Tuple{String,Int},Any}()
        function position(source_id::AbstractString, offset::Int)
            key = (String(source_id), offset)
            return get!(positions, key) do
                _julia_typed_source_position(authority, source_id, offset)
            end
        end

        for fixture in _julia_typed_source_object_rows("position_conversions")
            source_id = String(fixture["source_id"])
            offset = Int(fixture["offset"])
            value = position(source_id, offset)
            @test JuliaTypedSource.to_json(value) == Dict{String,Any}(
                "source_id" => source_id,
                "offset" => offset,
            )
            @test JuliaTypedSource.to_json(
                JuliaTypedSource.coordinates(
                    authority,
                    value;
                    context = JULIA_TYPED_SOURCE_CONTEXT,
                ),
            ) == Dict{String,Any}(
                "source_id" => source_id,
                "offset" => offset,
                "line" => fixture["line"],
                "column" => fixture["column"],
                "utf8_byte_offset" => fixture["utf8_byte_offset"],
            )
        end

        stable_position = positions[("unicode", 1)]
        detached_position = JuliaTypedSource.to_json(stable_position)
        detached_position["offset"] = 99
        @test detached_position["offset"] == 99
        @test JuliaTypedSource.to_json(stable_position)["offset"] == 1

        spans = Dict{String,Any}()
        for fixture in _julia_typed_source_object_rows("direct_spans")
            source_id = String(fixture["source_id"])
            start = Int(fixture["start"])
            stop = Int(fixture["end"])
            provenance = String(fixture["provenance"])
            span = JuliaTypedSource.direct_span(
                authority;
                start = position(source_id, start),
                stop = position(source_id, stop),
                provenance,
                context = JULIA_TYPED_SOURCE_CONTEXT,
            )
            spans[String(fixture["id"])] = span
            @test JuliaTypedSource.to_json(span) == Dict{String,Any}(
                "source_id" => source_id,
                "start" => start,
                "end" => stop,
                "provenance" => provenance,
            )
            @test JuliaTypedSource.materialize(
                authority,
                span;
                context = JULIA_TYPED_SOURCE_CONTEXT,
            ) == fixture["expected_text"]
        end

        for fixture in _julia_typed_source_object_rows("derived_text_cases")
            ordered_spans = Any[spans[String(id)] for id in fixture["span_ids"]]
            derived = JuliaTypedSource.derived_text(
                authority;
                policy = JuliaTypedSource.ConcatenateInOrder,
                spans = ordered_spans,
                context = JULIA_TYPED_SOURCE_CONTEXT,
            )
            @test JuliaTypedSource.to_json(derived) == Dict{String,Any}(
                "policy" => "concatenate_in_order",
                "spans" => Any[JuliaTypedSource.to_json(span) for span in ordered_spans],
            )
            @test JuliaTypedSource.materialize(
                authority,
                derived;
                context = JULIA_TYPED_SOURCE_CONTEXT,
            ) == fixture["expected_text"]
        end
    end

    @testset "authority ownership and four exact private diagnostics" begin
        sources = _julia_typed_source_decoded_sources()
        authority = JuliaTypedSource.SourceAuthority(sources = sources)
        sources["unicode"] = "changed"

        position(source_id::AbstractString, offset::Int) =
            _julia_typed_source_position(authority, source_id, offset)

        owned_span = JuliaTypedSource.direct_span(
            authority;
            start = position("unicode", 0),
            stop = position("unicode", 1),
            provenance = "input",
            context = JULIA_TYPED_SOURCE_CONTEXT,
        )
        @test JuliaTypedSource.materialize(
            authority,
            owned_span;
            context = JULIA_TYPED_SOURCE_CONTEXT,
        ) == "é"

        _julia_typed_source_expect_diagnostic(
            "source_mismatch",
            () -> JuliaTypedSource.direct_span(
                authority;
                start = position("unicode", 0),
                stop = position("ascii", 1),
                provenance = "input",
                context = JULIA_TYPED_SOURCE_CONTEXT,
            ),
            Dict{String,Any}(
                "rule_role" => "typed_source_fixture_rule",
                "invocation_role" => "typed_source_fixture_invocation",
                "source_id" => "unicode",
                "other_source_id" => "ascii",
            ),
        )
        _julia_typed_source_expect_diagnostic(
            "position_out_of_range",
            () -> position("unicode", 5),
            Dict{String,Any}(
                "rule_role" => "typed_source_fixture_rule",
                "invocation_role" => "typed_source_fixture_invocation",
                "source_id" => "unicode",
                "position_offset" => 5,
                "source_length" => 4,
            ),
        )
        _julia_typed_source_expect_diagnostic(
            "reversed_span",
            () -> JuliaTypedSource.direct_span(
                authority;
                start = position("unicode", 2),
                stop = position("unicode", 1),
                provenance = "capture",
                context = JULIA_TYPED_SOURCE_CONTEXT,
            ),
            Dict{String,Any}(
                "rule_role" => "typed_source_fixture_rule",
                "invocation_role" => "typed_source_fixture_invocation",
                "source_id" => "unicode",
                "start_offset" => 2,
                "end_offset" => 1,
            ),
        )

        foreign_authority = JuliaTypedSource.SourceAuthority(
            sources = Dict{String,String}("foreign" => "foreign text"),
        )
        foreign_span = JuliaTypedSource.direct_span(
            foreign_authority;
            start = _julia_typed_source_position(foreign_authority, "foreign", 0),
            stop = _julia_typed_source_position(foreign_authority, "foreign", 1),
            provenance = "input",
            context = JULIA_TYPED_SOURCE_CONTEXT,
        )
        _julia_typed_source_expect_diagnostic(
            "invalid_derived_provenance",
            () -> JuliaTypedSource.derived_text(
                authority;
                policy = JuliaTypedSource.ConcatenateInOrder,
                spans = Any[foreign_span],
                context = JULIA_TYPED_SOURCE_CONTEXT,
            ),
            Dict{String,Any}(
                "rule_role" => "typed_source_fixture_rule",
                "invocation_role" => "typed_source_fixture_invocation",
                "provenance_index" => 0,
                "source_id" => "foreign",
            ),
        )
    end

    if JULIA_TYPED_SOURCE_RED_MODE == "projection"
        @testset "detached exact 92 projection rows and seven aliases" begin
            engine = LinkedSpecRuntimeEngine(
                compile_spec(parse_spec("Top::\n I { return(\"ok\") }\n")),
            )
            expected_rows = JULIA_TYPED_SOURCE_CONTRACT["helper_projections"]
            rows = JULIA_TYPED_SOURCE_PROJECTION_ROWS(engine)
            @test rows == expected_rows

            families = JULIA_TYPED_SOURCE_CONTRACT["helper_projection_schema"]["families"]
            names = String[]
            for family in families
                append!(names, String(row[1]) for row in rows[String(family)])
            end
            @test length(names) == 92
            @test length(Set(names)) == 92

            expected_aliases = JULIA_TYPED_SOURCE_CONTRACT["compatibility_aliases"]
            @test JULIA_TYPED_SOURCE_COMPATIBILITY_ALIASES(engine) == expected_aliases
            @test length(expected_aliases) == 7

            rows["capture_mark"][1][2] = "wrong"
            @test JULIA_TYPED_SOURCE_PROJECTION_ROWS(engine) == expected_rows
        end

        @testset "mark capture cursor and alias results on all carriers" begin
            named_mark_contract = JSON3.read(
                read(
                    normpath(
                        joinpath(
                            @__DIR__,
                            "..",
                            "..",
                            "capability_conformance",
                            "complete_named_mark_contract.json",
                        ),
                    ),
                    String,
                ),
                Dict{String,Any},
            )
            fixture = named_mark_contract["fixture"]
            _julia_typed_source_assert_carriers(
                fixture["spec_source"],
                fixture["input"],
                "typed-source/complete-named-mark.spec",
                fixture["expected"],
            )
            _julia_typed_source_assert_carriers(
                JULIA_TYPED_SOURCE_CURSOR_SOURCE,
                "ab",
                "typed-source/cursor-control.spec",
                JULIA_TYPED_SOURCE_CURSOR_EXPECTED,
            )
            _julia_typed_source_assert_carriers(
                JULIA_TYPED_SOURCE_ALIAS_SOURCE,
                "é🙂  ab",
                "typed-source/compatibility-aliases.spec",
                JULIA_TYPED_SOURCE_ALIAS_EXPECTED,
            )
        end
    end
end

File: lua/test/typed_source_location_contract_test.lua
-- FUTURE-PARITY-BACKLOG.14.2.5.0.2 — dormant shared Lua typed source-location RED.
--
-- Ordinary Lua discovery is the explicit list in `tools/run_lua_local.sh`; that
-- list deliberately omits this pre-admission consumer. Run either mode through
-- repository-local project data from the repository root:
--
--   LINKEDSPEC_LUA_TYPED_SOURCE_RED_MODE=core \
--     bash tools/run_lua_project_data.sh puc lua/test/typed_source_location_contract_test.lua
--   LINKEDSPEC_LUA_TYPED_SOURCE_RED_MODE=core \
--     bash tools/run_lua_project_data.sh luajit lua/test/typed_source_location_contract_test.lua
--
--   LINKEDSPEC_LUA_TYPED_SOURCE_RED_MODE=projection \
--     bash tools/run_lua_project_data.sh puc lua/test/typed_source_location_contract_test.lua
--   LINKEDSPEC_LUA_TYPED_SOURCE_RED_MODE=projection \
--     bash tools/run_lua_project_data.sh luajit lua/test/typed_source_location_contract_test.lua
--
-- Core implementation makes `core` green. Projection implementation then
-- makes the independently nested `projection` mode green. Admission removes
-- this mode switch and adds the unchanged assertions to ordinary discovery.

local mode = os.getenv("LINKEDSPEC_LUA_TYPED_SOURCE_RED_MODE") or "core"
if mode ~= "core" and mode ~= "projection" then
  error(
    "LINKEDSPEC_LUA_TYPED_SOURCE_RED_MODE must be 'core' or 'projection', got '" ..
      tostring(mode) .. "'",
    0
  )
end

local json = require("linkedspec.json")

-- Core lookup is deliberately first. Convert only the ordinary absent-module
-- diagnostic to one stable RED boundary; propagate every other load failure.
local source_location_ok, source_location = pcall(require, "linkedspec.source_location")
if not source_location_ok then
  if tostring(source_location):match("module 'linkedspec%.source_location' not found") then
    error("Lua typed source RED: missing linkedspec.source_location", 0)
  end
  error(source_location, 0)
end

local linkedspec = require("linkedspec")

-- Projection lookup is strictly nested after the immutable core. Today both
-- modes stop at the missing module. Once the core lands, projection mode must
-- advance to these separately owned catalog APIs.
local typed_source_projection_rows
local typed_source_compatibility_aliases
if mode == "projection" then
  if type(linkedspec.typed_source_projection_rows) ~= "function" then
    error("Lua typed source RED: missing linkedspec.typed_source_projection_rows", 0)
  end
  if type(linkedspec.typed_source_compatibility_aliases) ~= "function" then
    error("Lua typed source RED: missing linkedspec.typed_source_compatibility_aliases", 0)
  end
  typed_source_projection_rows = linkedspec.typed_source_projection_rows
  typed_source_compatibility_aliases = linkedspec.typed_source_compatibility_aliases
end

local assertions = 0
local failures = {}

local function check(condition, label)
  assertions = assertions + 1
  if not condition then failures[#failures + 1] = label end
end

local function check_equal(actual, expected, label)
  check(
    actual == expected,
    (label or "values differ") .. ": expected " .. tostring(expected) .. ", got " .. tostring(actual)
  )
end

local function check_same_json(actual, expected, label)
  if actual == nil or expected == nil then
    check_equal(actual, expected, label)
    return
  end
  check_equal(json.encode(actual), json.encode(expected), label)
end

local function capture(operation)
  local ok, value = pcall(operation)
  return ok, value
end

local function read_file(path)
  local handle = assert(io.open(path, "rb"))
  local value = assert(handle:read("*a"))
  assert(handle:close())
  return value
end

local contract = json.decode(read_file("capability_conformance/typed_source_location_contract.json"))
local context = source_location.source_location_context({
  rule_role = "typed_source_fixture_rule",
  invocation_role = "typed_source_fixture_invocation",
})

local cursor_source = [[Top::
 /ab/ -> Done {
  after_match = cursor_pos()
  save_cursor()
  rewind_match_start()
  match_start = cursor_pos()
  restore_cursor()
  restored_match = cursor_pos()
  save_cursor()
  rewind_entry_start()
  entry_start = cursor_pos()
  restore_cursor()
  restored_entry = cursor_pos()
  return(hash(
   "after_match", after_match,
   "match_start", match_start,
   "restored_match", restored_match,
   "entry_start", entry_start,
   "restored_entry", restored_entry
  ))
 }

Done:
 /ab/
]]

local cursor_expected = json.harray({
  after_match = 2,
  entry_start = 0,
  match_start = 0,
  restored_entry = 2,
  restored_match = 2,
})

local alias_source = [[Top::OR{1,1}
 /(?<name>ab)/
 I { started = capture_slice_here() }
 E {
  return(array(
   started,
   capture_from_rule_start(),
   capture_len_from_rule_start(),
   capture_slice_length(),
   capture_rest_length(),
   entry_named_map(),
   match_named_map()
  ))
 }
]]

local alias_expected = json.array({
  json.null,
  "é🙂  ",
  4,
  4,
  6,
  json.harray({ name = "ab" }),
  json.harray({ name = "ab" }),
})

local function decoded_sources()
  local sources = json.harray()
  for _, fixture in ipairs(contract.sources) do sources[fixture.id] = fixture.decoded_text end
  return sources
end

local function position(authority, source_id, offset)
  return source_location.position(authority, {
    source_id = source_id,
    offset = offset,
    context = context,
  })
end

local function diagnostic_by_id(diagnostic_id)
  for _, diagnostic in ipairs(contract.diagnostics) do
    if diagnostic.id == diagnostic_id then return diagnostic end
  end
  error("missing typed source diagnostic fixture '" .. diagnostic_id .. "'", 0)
end

local function check_diagnostic(diagnostic_id, operation, expected_context)
  local ok, captured = capture(operation)
  check_equal(ok, false, diagnostic_id .. " rejects invalid value")
  if ok then return end

  check_equal(source_location.is_error(captured), true, diagnostic_id .. " error type")
  if not source_location.is_error(captured) then return end

  local diagnostic = diagnostic_by_id(diagnostic_id)
  local record = source_location.to_json(captured)
  check_equal(record.code, diagnostic.code, diagnostic_id .. " code")
  check_equal(record.phase, diagnostic.phase, diagnostic_id .. " phase")
  for _, field in ipairs(diagnostic.required_context) do
    check(record[field] ~= nil, diagnostic_id .. " required context " .. field)
  end
  for field, expected in pairs(expected_context) do
    check_equal(record[field], expected, diagnostic_id .. " context " .. field)
  end
  for _, forbidden in ipairs({
    "decoded_text",
    "source_text",
    "path",
    "match",
    "parser_state",
    "host_reference",
  }) do
    check_equal(record[forbidden], nil, diagnostic_id .. " private field " .. forbidden)
  end
end

local function compile_source(source)
  local parsed = linkedspec.parse_spec(source)
  linkedspec.validate_spec(parsed)
  return parsed, linkedspec.compile_spec(parsed)
end

local function check_carriers(source, input, identity, expected)
  local parsed, compiled = compile_source(source)
  local native = linkedspec.runtime_parse(linkedspec.runtime_engine(compiled), input).value

  local reconstructed_spec = linkedspec.spec_ast.from_json(
    "SpecFile",
    json.decode(json.encode(linkedspec.spec_ast.to_json(parsed)))
  )
  linkedspec.validate_spec(reconstructed_spec)
  local reconstructed = linkedspec.runtime_parse(
    linkedspec.runtime_engine(linkedspec.compile_spec(reconstructed_spec)),
    input
  ).value

  local generated = linkedspec.execute_generated_parser_v2(
    compiled,
    linkedspec.build_generated_rule_plan(compiled),
    input,
    identity
  )

  check_same_json(native, expected, identity .. " native")
  check_same_json(reconstructed, expected, identity .. " reconstructed")
  check_same_json(generated, expected, identity .. " generated plan")
end

check_equal(contract.contract_id, "linkedspec-typed-source-location-v1", "contract id")
check_equal(contract.expected_counts.sources, 3, "source fixture count")
check_equal(contract.expected_counts.position_conversions, 7, "position fixture count")
check_equal(contract.expected_counts.direct_spans, 6, "direct-span fixture count")
check_equal(contract.expected_counts.derived_text_cases, 3, "derived-text fixture count")
check_equal(source_location.node_type(context), "SourceLocationContext", "typed context")

do
  local sources = decoded_sources()
  local authority = source_location.source_authority({ sources = sources })
  check_equal(source_location.node_type(authority), "SourceAuthority", "typed source authority")

  local positions = {}
  local function fixture_position(source_id, offset)
    local key = source_id .. "\0" .. offset
    if positions[key] == nil then positions[key] = position(authority, source_id, offset) end
    return positions[key]
  end

  for _, fixture in ipairs(contract.position_conversions) do
    local value = fixture_position(fixture.source_id, fixture.offset)
    check_equal(source_location.node_type(value), "Position", fixture.id .. " type")
    check_same_json(
      source_location.to_json(value),
      json.harray({ source_id = fixture.source_id, offset = fixture.offset }),
      fixture.id .. " value"
    )
    check_same_json(
      source_location.coordinates(authority, value, { context = context }),
      json.harray({
        source_id = fixture.source_id,
        offset = fixture.offset,
        line = fixture.line,
        column = fixture.column,
        utf8_byte_offset = fixture.utf8_byte_offset,
      }),
      fixture.id .. " coordinates"
    )
  end

  local stable_position = fixture_position("unicode", 1)
  local detached_position = source_location.to_json(stable_position)
  detached_position.offset = 99
  check_equal(source_location.to_json(stable_position).offset, 1, "position record is detached")

  local spans = {}
  for _, fixture in ipairs(contract.direct_spans) do
    local span = source_location.direct_span(authority, {
      start = fixture_position(fixture.source_id, fixture.start),
      ["end"] = fixture_position(fixture.source_id, fixture["end"]),
      provenance = fixture.provenance,
      context = context,
    })
    spans[fixture.id] = span
    check_equal(source_location.node_type(span), "Span", fixture.id .. " type")
    check_same_json(
      source_location.to_json(span),
      json.harray({
        source_id = fixture.source_id,
        start = fixture.start,
        ["end"] = fixture["end"],
        provenance = fixture.provenance,
      }),
      fixture.id .. " value"
    )
    check_equal(
      source_location.materialize(authority, span, { context = context }),
      fixture.expected_text,
      fixture.id .. " materialized text"
    )
  end

  for _, fixture in ipairs(contract.derived_text_cases) do
    local ordered_spans = json.array()
    for index, span_id in ipairs(fixture.span_ids) do ordered_spans[index] = spans[span_id] end
    local derived = source_location.derived_text(authority, {
      policy = fixture.policy,
      spans = ordered_spans,
      context = context,
    })
    local expected_spans = json.array()
    for index, span in ipairs(ordered_spans) do expected_spans[index] = source_location.to_json(span) end
    check_equal(source_location.node_type(derived), "DerivedText", fixture.id .. " type")
    check_same_json(
      source_location.to_json(derived),
      json.harray({ policy = "concatenate_in_order", spans = expected_spans }),
      fixture.id .. " value"
    )
    check_equal(
      source_location.materialize(authority, derived, { context = context }),
      fixture.expected_text,
      fixture.id .. " materialized text"
    )
  end

  sources.unicode = "changed"
  local owned_span = source_location.direct_span(authority, {
    start = fixture_position("unicode", 0),
    ["end"] = fixture_position("unicode", 1),
    provenance = "input",
    context = context,
  })
  check_equal(
    source_location.materialize(authority, owned_span, { context = context }),
    "é",
    "authority snapshots decoded input"
  )

  check_diagnostic("source_mismatch", function()
    return source_location.direct_span(authority, {
      start = fixture_position("unicode", 0),
      ["end"] = fixture_position("ascii", 1),
      provenance = "input",
      context = context,
    })
  end, {
    rule_role = "typed_source_fixture_rule",
    invocation_role = "typed_source_fixture_invocation",
    source_id = "unicode",
    other_source_id = "ascii",
  })

  check_diagnostic("position_out_of_range", function()
    return position(authority, "unicode", 5)
  end, {
    rule_role = "typed_source_fixture_rule",
    invocation_role = "typed_source_fixture_invocation",
    source_id = "unicode",
    position_offset = 5,
    source_length = 4,
  })

  check_diagnostic("reversed_span", function()
    return source_location.direct_span(authority, {
      start = fixture_position("unicode", 2),
      ["end"] = fixture_position("unicode", 1),
      provenance = "capture",
      context = context,
    })
  end, {
    rule_role = "typed_source_fixture_rule",
    invocation_role = "typed_source_fixture_invocation",
    source_id = "unicode",
    start_offset = 2,
    end_offset = 1,
  })

  local foreign_authority = source_location.source_authority({
    sources = json.harray({ foreign = "foreign text" }),
  })
  local foreign_span = source_location.direct_span(foreign_authority, {
    start = position(foreign_authority, "foreign", 0),
    ["end"] = position(foreign_authority, "foreign", 1),
    provenance = "input",
    context = context,
  })
  check_diagnostic("invalid_derived_provenance", function()
    return source_location.derived_text(authority, {
      policy = "concatenate_in_order",
      spans = json.array({ foreign_span }),
      context = context,
    })
  end, {
    rule_role = "typed_source_fixture_rule",
    invocation_role = "typed_source_fixture_invocation",
    provenance_index = 0,
    source_id = "foreign",
  })
end

if mode == "projection" then
  local _, catalog_compiled = compile_source("Top::\n I { return(\"ok\") }\n")
  local catalog_engine = linkedspec.runtime_engine(catalog_compiled)
  local rows = typed_source_projection_rows(catalog_engine)
  check_same_json(rows, contract.helper_projections, "exact detached 92 projection rows")

  local names = {}
  local name_count = 0
  for _, family in ipairs(contract.helper_projection_schema.families) do
    for _, row in ipairs(rows[family]) do
      name_count = name_count + 1
      check_equal(names[row[1]], nil, row[1] .. " projection name is unique")
      names[row[1]] = true
    end
  end
  check_equal(name_count, 92, "projection row count")

  local aliases = typed_source_compatibility_aliases(catalog_engine)
  check_same_json(aliases, contract.compatibility_aliases, "exact detached seven compatibility aliases")
  check_equal(#aliases, 7, "compatibility alias count")

  rows.capture_mark[1][2] = "wrong"
  aliases[1][2] = "wrong"
  check_same_json(
    typed_source_projection_rows(catalog_engine),
    contract.helper_projections,
    "projection rows are detached"
  )
  check_same_json(
    typed_source_compatibility_aliases(catalog_engine),
    contract.compatibility_aliases,
    "compatibility aliases are detached"
  )

  local named_mark_contract = json.decode(
    read_file("capability_conformance/complete_named_mark_contract.json")
  )
  local fixture = named_mark_contract.fixture
  check_carriers(
    fixture.spec_source,
    fixture.input,
    "typed-source/complete-named-mark.spec",
    fixture.expected
  )
  check_carriers(
    cursor_source,
    "ab",
    "typed-source/cursor-control.spec",
    cursor_expected
  )
  check_carriers(
    alias_source,
    "é🙂  ab",
    "typed-source/compatibility-aliases.spec",
    alias_expected
  )
end

if #failures == 0 then
  io.stdout:write(
    "Lua dormant typed source-location ", mode, ": ", assertions, " assertions passed\n"
  )
else
  io.stderr:write(
    "Lua dormant typed source-location ", mode, ": ", #failures, " of ", assertions,
    " assertions failed\n"
  )
  for _, message in ipairs(failures) do io.stderr:write("- ", message, "\n") end
  os.exit(1)
end
